### ISSUE TITLE
Mypy shows object for values of large dictionaries

### ISSUE THREAD

**Issue by PythonCoderAS** [python/mypy#12831]
**Bug Report**

I have a large block of decently-nested JSON that I wrote types for using TypedDicts. I would like to verify that my types are correct by assigning a Python representation of one of the JSON blobs using the appropriate type variable. However, I am getting errors like

```
tests/test.py:15: error: Incompatible types in assignment (expression has type "Dict[str, List[List[object]]]", variable has type "SeriesVolumeCovers")
```

However, the value I am assigning is

```py
{
    "covers": [
        [
            1,
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/1/1.png",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/1/1.webp",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/1/1_blur.png",
        ],
        [
            2,
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/2/2.png",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/2/2.webp",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/2/2_blur.png",
        ],
    ]
}
```

My types look like this: 

```py
SeriesVolumeCover = List[Union[str, int]]

class SeriesVolumeCovers(TypedDict):
    covers: List[SeriesVolumeCover]
```

The two should match, but for some reason mypy is complaining.

**To Reproduce**

1. Create types like so:

```py
from typing import List, Union, TypedDict

SeriesVolumeCover = List[Union[str, int]]

class SeriesVolumeCovers(TypedDict):
    covers: List[SeriesVolumeCover]
```

2. Assign variable

```py
data: SeriesVolumeCovers = {
    "covers": [
        [
            1,
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/1/1.png",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/1/1.webp",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/1/1_blur.png",
        ],
        [
            2,
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/2/2.png",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/2/2.webp",
            "/media/manga/Kaguya-Wants-To-Be-Confessed-To/volume_covers/2/2_blur.png",
        ],
    ]
}
```

**Expected Behavior**

I expected the types to match.

**Actual Behavior**

The types did not match.

**Your Environment**

- Mypy version used: 0.950
- Mypy command-line flags: No flags
- Mypy configuration options from `mypy.ini` (and other config files): No mypy.ini
- Python version used: 3.9
- Operating system and version: MacOS Big Sur/Ubuntu 20 LTS

**Comment by Zeckie** [comment]
I tried to reproduce the issue on mypy playground, but needed to make a change (added an intermediate variable) to get the same error.

https://mypy-play.net/?mypy=0.950&python=3.9&gist=78d258e683165265cd6dada6dbee741d

This looks to me as though it is a known issue - https://github.com/python/mypy/labels/topic-join-v-union (eg.  https://github.com/python/mypy/issues/12056 )

**Comment by PythonCoderAS** [comment]
Oh yea in my actual test setup I did have an intermediate variable, as the python structure files were automatically generated

**Comment by JelleZijlstra** [comment]
Agree, this is another join-vs-union, we have plenty of open issues about that.
